### ISSUE TITLE
Password Grant Not Following Spec Properly

### ISSUE THREAD

**Issue by jasonlewis** [thephpleague/oauth2-server#153]
I understand that the spec can sometimes be up to different interpretations but I'm going to go ahead with this anyway just so I can get some other input on this.

I'll be referring to parts from [section 4.3](http://tools.ietf.org/html/rfc6749#section-4.3) and [section 4.3.2](http://tools.ietf.org/html/rfc6749#section-4.3.2) of the OAuth2 spec.

Initially the spec states that:

>  The client requests an access token from the authorization server's token endpoint by including the credentials received from the resource owner.  **When making the request, the client authenticates with the authorization server.**

Which seems to indicate that the client should always be authenticated with the server. It then goes on to say:

> require client authentication for confidential clients or for any client that was issued client credentials (or with other authentication requirements)

And then, more importantly:

> authenticate the client if client authentication is included

From this I think it's safe to say that for the password grant type the client does not need to supply authentication.

The current implementation of this requires that client authentication be supplied. Just wanting some thoughts on this and wondering if it should be changed so clients are not required to authenticate.

**Comment by Danack** [comment]
Hi Alex,

Apologies for opening this issue again, however I'm pretty sure how this library has implemented the password_grant type is not correct, and is making the easy path be to inadvertently expose the client_secret.

I'm trying to use the library to allow a webpage app to authenticate users. Section 2.1 describes the types of client https://tools.ietf.org/html/rfc6749#section-2.1 :

> OAuth defines two client types, based on their ability to
>    authenticate securely with the authorization server (i.e., ability to
>    maintain the confidentiality of their client credentials):
> 
>    confidential
>       Clients capable of maintaining the confidentiality of their
>       credentials (e.g., client implemented on a secure server with
>       restricted access to the client credentials), or capable of secure
>       client authentication using other means.
> 
>    public
>       Clients incapable of maintaining the confidentiality of their
>       credentials (e.g., clients executing on the device used by the
>       resource owner, such as an installed native application or a web
>       browser-based application), and incapable of secure client
>       authentication via any other means.

As it is a webpage app, it is not possible for it to maintain the confidentiality of their credentials, and so should be considered a public app.

Also from section 2.1, although I might also want to use the Oauth library in a 'confidential' way elsewhere, those should be considered separate components.

>  A client may be implemented as a distributed set of components, each
>    with a different client type and security context (e.g., a
>    distributed client with both a confidential server-based component
>    and a public browser-based component).  If the authorization server
>    does not provide support for such clients or does not provide
>    guidance with regard to their registration, the client SHOULD
>    register each component as a separate client.

   
Those two pieces of information affect the meaning of  https://tools.ietf.org/html/rfc6749#section-4.3.2
   
>   If the client type is confidential or the client was issued client
>    credentials (or assigned other authentication requirements), the
>    client MUST authenticate with the authorization server as described
>    in Section 3.2.1.

The webpage client is not confidential (it is public) and any secure client should have registered any secure component as a separate client. 

tl:dr the password_grant type shouldn't need the client_secret to authenticate for webpage applications.

Although copy+pasta+modifying the PasswordGrant type, and registering it as a custom type is trivial, that really ought to be mentioned on the [password grant page](https://oauth2.thephpleague.com/authorization-server/resource-owner-password-credentials-grant/)

I've opened a PR to modify the text on that page, but also I think PasswordGrantWithoutClientSecret should probably be something this library offers out of the box.
